PP-1210: Expect function should validate server_state when scheduling is turned off

Mahalty opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

Issue ID:

- **[PP-1210](https://pbspro.atlassian.net/browse/PP-1210)**

#### Affected Platform(s) and PBS Version
* Platform: All
* version where observed: 18.0

#### Cause / Analysis / Design

- if a job is submitted after scheduling is turned off, there are chances that the job might get considered for the running scheduling cycle. 

#### Solution Description
* In order to handle the above mentioned race condition, PTL needs to Validate 'server_state' to be 'Idle' whenever scheduling is turned off along with verifying the scheduling attribute in qstat -Bf command

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

hirenvadalia commented:
@Mahalty After looking in to ticket description I think there is some race condition in PBS itself like jobs are running even scheduling is off. If that is true then why we are handling in PTL?

agrawalravi90 commented:
@bhroam is the best person to answer this as he explained this to me a few days back, but I'll try to answer: Setting the server's 'scheduling' attribute to False is intended to stop any further scheduling cycles from happening, but if there's one happening right now, it is not supposed to kill it. And that's what causes the "race condition". I think this is the intended behavior from PBSPro's side (if not then ya, we should fix PBSPro instead of PTL). From PTL's standpoint, if there's a need to "turn off scheduling", then we should either 1) set scheduling to False and wait for any currently running sched cycles to complete (which i believe is what this PR is proposing), or 2) just kill the scheduler daemon. I vote for the option 1 simply because killing the scheduler might lead to undefined behavior.

hirenvadalia commented:
@agrawalravi90 Thanks for explanation, I understand your point but I am still not clear about race condition vs intended behavior part. So I will wait for @bhroam's reply on this.

bhroam commented:
@hirenvadalia as @agrawalravi90 said, turning scheduling off will just stop future cycles from happening.  If the scheduler is currently in cycle, it will complete.  If you want to make sure no jobs will start after you set scheduling to false, you need to do an expect() on the server state.  The server state will be "scheduling" while the scheduler is in cycle.  If you wait for the state to become idle, you should be safe.  Of course this situation changes when multisched checks in.  The server's state only talks about the default scheduler.  Each scheduler has its own state.

So this is expected behavior.  It is not a race condition.

agrawalravi90 commented:
Oh, I wasn't aware of the multi-sched behavior changes, we should probably create a ticket to capture that for PTL.

hirenvadalia commented:
@bhroam Thanks for explanation. So if I understand this expected behavior correctly then job which is submitted after scheduling off will considered only if it get submitted before sched does query_jobs and after scheduling off but if job submitted after sched has already completed query_jobs then that job will not get considered, right?

Anyway, if it is expected behavior then why we are putting workaround in test framework?
If we have to do changes in test framework then I don't think Server.manager or Server.expect is right place to do changes, we should introduce new method to Server (or Sched) (say set_scheduling) with flag then based on flag's value, set_scheduling should wait for server to become idle. This will allow user to write test case in which they want don't want to wait for server to become idle and it will also handle multisched case if we put set_scheduling in Sched class. Why you think?

bhroam commented:
I think the framework should handle this case.  I view it as similar to when we do an expect() for job_state=R, we do also check for substate=42.  It helps the test writer.  From an OOP point of view you're the new scheduler method makes sense.  Whether or not a scheduler runs cycles is a property of the scheduler itself.  In the same vein we could also introduce a new method for running a scheduling cycle.  It could set scheduling to true and then wait for the cycle to finish.  In any case, that isn't part of this PR.

agrawalravi90 commented:
My 2 cents: if most tests assume that "setting 'scheduling' to False" equates "no further jobs will be run", then it might be a good idea to have the expect for setting scheduling=False check whether server_state is idle. If, however, existing PTL tests don't make this assumption, then ya, I agree that we should have a special interface, maybe server.turn_scheduling_off(), which would set scheduling to False and check for server_state=Idle.

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -6641,12 +6642,23 @@ def manager(self, cmd, obj_type, attrib=None, id=None, extend=None,
+            # If scheduling is set to false then check for
+            # 'server_state' to be 'Idle'
+            if attrib and 'scheduling' in attrib.keys():
+                if str(attrib['scheduling']) in ('False',
```
Please make use of PTL_FALSE instead tuple of all possible value for False.

agrawalravi90 commented:
Thanks for making the changes @Mahalty .

I'm confused by the title of the bug. It says "Expect function should validate server_state when scheduling is turned off", that's not a bug, that sounds like the solution to the bug. The bug seems to be a race condition, which I believe was the original title of the ticket, which seems like the correct title to me. Could you please explain the rationale behind updating the ticket title?

Also, please post  "before bugfix" and "after bugfix" logs so that reviewers can see how the bug was resolved. Thanks!

Mahalty commented:
@hirenvadalia  I have addressed your review comment. Please review the latest commit.
@agrawalravi90  As part of this ticket I am not fixing the test's race condition, rather I am fixing the PTL framework (as per the discussion we had) to handle the known race condition in PBS. With this fix PTL framework will make sure that the PBS is consistent(scheduling and server_state dependency)

agrawalravi90 commented:
Ok, I just noticed that this was not captured by a PTL test, it is more of an enhancement to PTL framework to handle possible use cases in the future. I'm ok with the changes.

mike0042 commented:
Running check: 'checkpep8' ... FAILED
test/fw/ptl/lib/pbs_testlib.py:6670:80: E501 line too long (85 > 79 characters)
            if attrib and isinstance(attrib, dict) and 'scheduling' in attrib.keys():

Mahalty commented:
@mike0042 I have run pep8 on my code and also rebased my branch. Please take this PR forward.
